### ISSUE TITLE
mingw issue with executable path

### ISSUE THREAD

**Issue by ambitslix** [ianlancetaylor/libbacktrace#42]
Getting  the following error from callback after backtrace_full:

```
static void
ErrorCallback(void* data, const char* message, int err)

executable file is not COFF (0)
failed to read executable information (-1)
```
Then it it's an infinite loop of last error. Under Linux it runs fine.

Environment:

Statically linked executable
ExceptionTest.exe: PE32+ executable (console) x86-64, for MS Windows
OS:Ubuntu/Wine
MXE static compiler

Any clues?

**Comment by ianlancetaylor** [comment]
What are the first 20 bytes of the file?

**Comment by ambitslix** [comment]
![Screenshot from 2020-05-16 19-53-20](https://user-images.githubusercontent.com/3694590/82134677-00665980-97af-11ea-8de8-0530089eb8d1.png)

Let me know, and I will send you the executable if that helps.

**Comment by ianlancetaylor** [comment]
Thanks.  The start of that file looks fine.  I don't see how it could produce that error.

Can you run it under the debugger and see what is happening?  The error is being returned by the function `coff_add` in pecoff.c.

**Comment by ambitslix** [comment]
I ran it under the debugger, but it had errors before it got to that error, so I decided to reduce the complexity and used an example program that has minimal complexity, now I'm just seeing no trace but no errors. Is there a simple executable from the library source to test with? Otherwise I will create a barebones C program to further test this.

**Comment by matanui159** [comment]
I did a bit of a code dive and the problem seems to be that `/proc/self/exe` is the first valid executable filename, which just happens to point to the wine process instead of the current process.

**Comment by matanui159** [comment]
Some more notes before i head to bed.
- @ambitslix seems to mention that it also happens on windows which would invalidate this theory, but this is atleast what is causing it on linux under wine
- I had a play around with using [GetModuleFileNameA](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea) but it returns windows-like paths (`Z:\home\...`) which may work under windows but not under wine (if that even is important)
- still not entirely sure why linux-paths are working but not windows paths. do the posix functions on windows act the same? is it just wine oddities?

**Comment by ambitslix** [comment]
@matanui159 yes there are some of the same issues under msys2(Windows) however the executable path was definitely one of the problems under wine in Linux. But to clarify it didn't had to do with GetModuleFileName. It had to do with the variable passed to **backtrace_create_state** which in my case was temporary for some reason. Now both argv[0] and GetModuleFileName work under both wine and msys2 under Windows. They both return the same path under wine and msys2 for me, and they're windows paths. So that doesn't seem to be the problem anymore, and wasn't the problem in the case of btest.c.

1. So backtrace_state should copy the exec path passed to backtrace_create_state for one thing. I tried to debug btest.c in msys2 but could not yet find why there is no backtrace.

2. Also tests do not build under cross-compiler with make check

**Comment by matanui159** [comment]
Also as a side note I noticed that if it fails to open the provided filename it falls back to alternatives. It would be nice if it failed with an error straight away if it couldn't open the provided filename since it would make issues like this easier to debug.

**Comment by ambitslix** [comment]
Yes I agree @matanui159. I'm opening another issue for the remaining problem which is that with mingw the library doesn't work.

**Comment by matanui159** [comment]
I'll probably keep looking into that today although it will be slightly biased since I'm running it under wine on Linux. I really wanna get this project working, it's a really neat little library.

**Comment by krlmlr** [comment]
FWIW, whereami might be a more robust way to automatically determine the executable file name. This also works on MSYS2/Windows: https://github.com/gpakosz/whereami.

**Comment by lurobi** [comment]
Commenting on this to raise awareness, and see if a work-around can be identified.  I'm using Gfortran 9.2 on msys2/mingw64 and it exhibits what may be the same problem.  In my case when a segfault is issued, gfortran runtime uses libbacktrace to print the stack.  The result I see on windows msys2/mingw64 is the following:

```
Error termination. Backtrace:

Could not print backtrace: libbacktrace could not find executable to open
#0  0xffffffff
#1  0xffffffff
#2  0xffffffff
....
#19  0xffffffff
```

**Comment by ianlancetaylor** [comment]
This was most likely fixed by c1c86fa2.  Please comment if it is still a problem.  Thanks.
